File: Android/Java/Telemedicine/app/src/main/java/com/telemedicine/activity/auth/LoginActivity.java
package com.telemedicine.activity.auth;

import androidx.appcompat.app.AppCompatActivity;

import android.app.ProgressDialog;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.telemedicine.API;
import com.telemedicine.APIClient;
import com.telemedicine.R;
import com.telemedicine.Utils;
import com.telemedicine.activity.admin.AdminActivity;
import com.telemedicine.activity.dokter.DokterActivity;
import com.telemedicine.activity.pasien.PasienActivity;
import com.telemedicine.model.LoginResponse;
import com.telemedicine.model.RegisterResponse;
import com.telemedicine.model.User;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LoginActivity extends AppCompatActivity {
    TextView register;
    EditText email, password;
    Button login;
    ProgressDialog progressDialog;
    API api;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        register = findViewById(R.id.tv_register);
        email = findViewById(R.id.et_email);
        password = findViewById(R.id.et_password);
        login = findViewById(R.id.btn_login);
        progressDialog = Utils.PROGRESS(this);
        api = APIClient.getClient().create(API.class);
        SharedPreferences preferences= getSharedPreferences("auth", 0);
        if(preferences.getInt("id",-1)!=-1){
            gotoDashboard(preferences.getInt("id",-1));
            finish();
        }
        register.setOnClickListener(view -> {
            startActivity(new Intent(this, RegisterActivity.class));
            finish();
        });
        login.setOnClickListener(view -> login());

    }
    private void login(){
        progressDialog.show();
        Call<LoginResponse> loginCall = api.login(email.getText().toString(), password.getText().toString());
        loginCall.enqueue(new Callback<LoginResponse>() {
            @Override
            public void onResponse(Call<LoginResponse> call, Response<LoginResponse> response) {
                progressDialog.dismiss();
                if(response.body().id!=-1){
                    SharedPreferences.Editor editor = getSharedPreferences("auth", 0).edit();
                    editor.putInt("id", response.body().id);
                    editor.apply();
                    gotoDashboard(response.body().id);
                    finish();
                }else{
                    Toast.makeText(LoginActivity.this,"Email atau password salah",Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<LoginResponse> call, Throwable t) {
                progressDialog.dismiss();
                Toast.makeText(LoginActivity.this, t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });
    }
    private void gotoDashboard(int id){
        Call<User> getUserCall = api.getUser(id);
        getUserCall.enqueue(new Callback<User>() {
            @Override
            public void onResponse(Call<User> call, Response<User> response) {
                if(response.body().role.equals("Admin")){
                    startActivity(new Intent(LoginActivity.this, AdminActivity.class));
                    finish();
                }else if(response.body().role.equals("Dokter")){
                    startActivity(new Intent(LoginActivity.this, DokterActivity.class));
                    finish();
                }else{
                    startActivity(new Intent(LoginActivity.this, PasienActivity.class));
                    finish();
                }
                progressDialog.dismiss();
            }
            @Override
            public void onFailure(Call<User> call, Throwable t) {
                progressDialog.dismiss();
                finish();
            }
        });
    }
}

File: Android/Kotlin/StoryApp/app/src/main/java/com/dicoding/picodiploma/storyapp/LoginActivity.kt
package com.dicoding.picodiploma.storyapp

import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Context
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.view.*
import android.widget.Toast
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.preferencesDataStore
import androidx.lifecycle.ViewModelProvider
import com.dicoding.picodiploma.storyapp.databinding.ActivityLoginBinding
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "settings")
class LoginActivity : AppCompatActivity() {

    private lateinit var binding: ActivityLoginBinding
    private lateinit var myEditText: MyEditText
    private lateinit var loginViewModel: AllViewModel

    companion object{
        private const val TAG = "Login Activity"
    }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)

        myEditText = findViewById(R.id.passwordEditText)

        setupViewModel()
        playAnimation()
        binding.loginButton.setOnClickListener {
            val inputEmail = binding.emailEditText.text.toString()
            val inputPassword = binding.passwordEditText.text.toString()

            login(inputEmail, inputPassword)
        }
        binding.regisButton.setOnClickListener {
            val intent = Intent(this, RegisterActivity::class.java)
            startActivity(intent)
        }
     
    }
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.option_menu, menu)

        val logoutSetting = menu.findItem(R.id.menu_logout)

        logoutSetting.isVisible = false

        return true
    }
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when(item.itemId) {
            R.id.menu_language -> {
                val intent = Intent(Settings.ACTION_LOCALE_SETTINGS)
                startActivity(intent)
                return true
            }
        }
        return true
    }

    private fun login(inputEmail: String, inputPassword: String) {
        val client = ApiConfig.getApiService().loginUser(inputEmail, inputPassword)
        client.enqueue(object: Callback<LoginResponse> {
            override fun onResponse(call: Call<LoginResponse>, response: Response<LoginResponse>) {
                val responseBody = response.body()
                Log.d(TAG, "onResponse: $responseBody")
                if (response.isSuccessful && responseBody?.message == "success") {
                    loginViewModel.saveUser(UserData(responseBody.loginResult.token, true))
                    Toast.makeText(
                        this@LoginActivity,
                        getString(R.string.loginsuccess),
                        Toast.LENGTH_SHORT
                    ).show()
                    val intent = Intent(this@LoginActivity, MainActivity::class.java)
                    startActivity(intent)
                } else {
                    Log.e(TAG, "onFailure1: ${response.message()}")
                    Toast.makeText(
                        this@LoginActivity, getString(R.string.loginfail),
                        Toast.LENGTH_SHORT
                    ).show()
                }
            }

            override fun onFailure(call: Call<LoginResponse>, t: Throwable) {
                Log.e(TAG, "onFailure2: ${t.message}")
                Toast.makeText(
                    this@LoginActivity,
                    getString(R.string.loginfail),
                    Toast.LENGTH_SHORT
                ).show()

            }
        })
    }

    private fun setupViewModel(){
        loginViewModel = ViewModelProvider(
            this,
            ViewModelFactory(UserPreference.getInstance(dataStore)))[AllViewModel::class.java]
        loginViewModel.getUser().observe(this) { user ->
            if(user.isLogin) {
                val intent = Intent(this, MainActivity::class.java)
                startActivity(intent)
                finish()
            }
        }
    }
    private fun playAnimation(){
        val emailTextView = ObjectAnimator.ofFloat(binding.emailTextView, View.ALPHA, 1f).setDuration(500)
        val emailEditTextLayout = ObjectAnimator.ofFloat(binding.emailEditTextLayout, View.ALPHA, 1f).setDuration(500)
        val passwordTextView = ObjectAnimator.ofFloat(binding.passwordTextView, View.ALPHA, 1f).setDuration(500)
        val passwordEditTextLayout = ObjectAnimator.ofFloat(binding.passwordEditTextLayout, View.ALPHA, 1f).setDuration(500)
        val login = ObjectAnimator.ofFloat(binding.loginButton, View.ALPHA, 1f).setDuration(500)
        val register = ObjectAnimator.ofFloat(binding.regisButton, View.ALPHA, 1f).setDuration(500)

        AnimatorSet().apply {
            playSequentially( emailTextView, emailEditTextLayout, passwordTextView, passwordEditTextLayout, login, register)
            startDelay = 500
        }.start()


       
        
    }
}